[LoopInterchange] Remove 'S' Scalar Dependencies

sjoerdmeijer opened the pull request:
We are not handling 'S' scalar dependencies correctly and have at least
the following miscompiles related to that:

[LoopInterchange] incorrect handling of scalar dependencies and dependence vectors starting with ">" #54176
[LoopInterchange] Interchange breaks program correctness #46867
[LoopInterchange] Loops should not interchanged due to dependencies #47259
[LoopInterchange] Loops should not interchanged due to control flow #47401

 This patch does no longer insert the "S" dependency/direction into the
 dependency matrix, so a dependency is never "S". We seem to have
 forgotten what the exact meaning is of this dependency type, and don't
 see why it should be treated differently.

 We prefer correctness over incorrect and more aggressive results. I.e.,
 this prevents the miscompiles at the expense of handling less cases,
 i.e. making interchange more pessimistic. However, some of the cases
 that are now rejected for dependence analysis reasons, were rejected
 before too but for other reasons (e.g. profitability). So at least for
 the llvm regression tests, the number of regression are very reasonable.
 This should be a stopgap. We would like to get interchange enabled by
 default and thus prefer correctness over unsafe transforms, and later
 see if we can get solve the regressions.

kasuga-fj commented on llvm/lib/Transforms/Scalar/LoopInterchange.cpp:
```diff
@@ -187,11 +187,19 @@ static void interChangeDependencies(CharMatrix &DepMatrix, unsigned FromIndx,
 // if the direction matrix, after the same permutation is applied to its
 // columns, has no ">" direction as the leftmost non-"=" direction in any row.
 static bool isLexicographicallyPositive(std::vector<char> &DV) {
+  bool HasScalar = false;
```
nit: unused?

kasuga-fj commented on llvm/lib/Transforms/Scalar/LoopInterchange.cpp:
```diff
@@ -187,11 +187,19 @@ static void interChangeDependencies(CharMatrix &DepMatrix, unsigned FromIndx,
 // columns, has no ">" direction as the leftmost non-"=" direction in any row.
 static bool isLexicographicallyPositive(std::vector<char> &DV) {
+  bool HasScalar = false;
+  bool FirstPosScalar = true;
```
I don't fully understand what `FirstPosScalar` means. This legality check seems to allow the permutation of `[S <]` to `[< S]`. Is this transformation legal? Or can such a case not happen? I tried some loops that have a row of `[S <]`, and as far as these cases, the legality check rejects the interchange because of another row like `[S =]`.

CongzheUalberta replied:
I would second this. It looks to me that this patch allows some specific pattern to be interchanged without a schematic reasoning.

Meinersbur commented:
I honestly never understood why a dependency that is "scalar" (that is, the dimension is not referenced in the index expression) is handled fundamentally different. If not references it should be equivalent to accessing (any of) the previous iteration. How about we remove 
```
          if (D->isScalar(II)) {
            Direction = 'S';
            Dep.push_back(Direction);
```
in populateDependencyMatrix entirely, so a dependency is never `S`?

In any case, the patch in the current form should be conservatively correct.

sjoerdmeijer commented:
> I honestly never understood why a dependency that is "scalar" (that is, the dimension is not referenced in the index expression) is handled fundamentally different. If not references it should be equivalent to accessing (any of) the previous iteration. How about we remove
> 
> ```
>           if (D->isScalar(II)) {
>             Direction = 'S';
>             Dep.push_back(Direction);
> ```
> 
> in populateDependencyMatrix entirely, so a dependency is never `S`?
> 
> In any case, the patch in the current form should be conservatively correct.

I would be happy to remove it in the way you suggested, @Meinersbur. My only minor concern is that if we want to deal with these sort of dependencies later, we would need to restore some of this logic, so removing it now would mean more churn later? But let me know what you prefer, I am happy either way (leave it as it is because it is conservative/correct as you say, or implement your suggestion).

kasuga-fj commented:
I agree with removing the 'S'. I think leaving the incorrect dependencies risks causing correctness issues, especially when we change the code in the future.  At the extreme, we need to insert something like `if (Direction == 'S') { ... }` as this patch does when adding new functions.

Meinersbur commented:
@sjoerdmeijer What kind of interchange do you want to preserve with `S`? IMHO the current code doesn't even understand what it should mean (or it was forgotten about at some point), so if we want to re-add it, it should get another full review.

Besides, this PR currently will just reject everything with `S`, but `DependenceInfo` may be able to determine the direction already.

sjoerdmeijer commented:
> @sjoerdmeijer What kind of interchange do you want to preserve with `S`? IMHO the current code doesn't even understand what it should mean (or it was forgotten about at some point), so if we want to re-add it, it should get another full review.
> 
> Besides, this PR currently will just reject everything with `S`, but `DependenceInfo` may be able to determine the direction already.

Okay, excellent points. I will prepare a new revision that removes `S`.

sjoerdmeijer commented:
Thanks for the review @Meinersbur , @kasuga-fj: this now no longer inserts 'S' into the dependency matrix, it should default to '*'.

kasuga-fj commented on llvm/test/Transforms/LoopInterchange/inner-only-reductions.ll:
```diff
@@ -18,7 +18,7 @@
 
 ; CHECK: --- !Missed
 ; CHECK-NEXT: Pass:            loop-interchange
-; CHECK-NEXT: Name:            UnsupportedPHI
+; CHECK-NEXT: Name:            Dependence
```
I believe it is important for this test to be rejected by "UnsupportedPHI"

sjoerdmeijer replied:
Can you elaborate why it is important?
Very simplified, the structure and the order of the checks is as follows:

    canInterchangeLoops() {
        isLegalToInterChangeLoops();
        currentLimitations();
    }

where `currentLimitations()` is responsible for rejecting the `UnsupportedPHI`, and `isLegalToInterChangeLoops()` when dependencies are getting in the way.

Because we reject S dependencies and because of this order of checks, this case and the one below will now first be rejected for dependencies reasons. I might be missing something, but I don't see how it matters.

kasuga-fj replied:
It looks to me like this test checks the behavior of `currentLimitations`. If so, we need to prepare a case that passes `isLegalToInterChangeLoops`, but is rejected by `currentLimitations`. I may be wrong, or it may no longer be possible to make such cases.

sjoerdmeijer replied:
Ok, I see what you mean. I think this is mostly/somewhat mitigated with the xfails, and there is still a test that tests part of `currentLimitations` in `multilevel-partial-reduction.ll` e.g.

kasuga-fj replied:
I see, thanks. For now, I think there are two perspectives:

- This loop should be rejected by LoopInterchange for some reason.
- After scalar dependencies are handled correctly, this one should be rejected by `currentLimitations`.

IMHO, to guarantee the first point, it's better to take your current change and add some comments to describe the context of the second point, rather than XFAILS.

sjoerdmeijer replied:
Thanks, have added a comment.

kasuga-fj commented on llvm/test/Transforms/LoopInterchange/reductions-across-inner-and-outer-loop.ll:
```diff
@@ -153,7 +153,7 @@ for1.loopexit:                                 ; preds = %for1.inc
 ; REMARKS: --- !Missed
 ; REMARKS-NEXT: Pass:            loop-interchange
-; REMARKS-NEXT: Name:            UnsupportedPHIOuter
+; REMARKS-NEXT: Name:            Dependence
```
As with `inner-only-reductions.ll`, I think the reason for failure is important for this test.

kasuga-fj commented on llvm/test/Transforms/LoopInterchange/innermost-latch-uses-values-in-middle-header.ll:
```diff
@@ -18,35 +18,28 @@ define void @innermost_latch_uses_values_in_middle_header() {
-; CHECK-NEXT:    [[INDVAR_INNERMOST:%.*]] = phi i64 [ [[TMP1:%.*]], %[[INNERMOST_LATCH_SPLIT:.*]] ], [ 4, %[[INNERMOST_HEADER_PREHEADER]] ]
-; CHECK-NEXT:    br label %[[MIDDLE_HEADER_PREHEADER]]
+; CHECK-NEXT:    [[INDVAR_INNERMOST:%.*]] = phi i64 [ [[TMP3:%.*]], %[[INNERMOST_LATCH:.*]] ], [ 4, %[[MIDDLE_HEADER]] ]
+; CHECK-NEXT:    br label %[[INNERMOST_BODY:.*]]
```
These loops do not seem to be interchanged. It seems that a Basic Block named `XXX.split` is generated when the loops are interchanged. So it is possible that the same thing is happening in the tests where the string `XXX_SPLIT` is removed.

kasuga-fj reviewed (commented):
The implementation looks good to me. but some tests do not appear to have achieved their original objectives. I'm not sure what we should do in such cases, but in my opinion a mechanical update does not make sense. Marking `XFAIL` may be one option.

sjoerdmeijer commented on llvm/test/Transforms/LoopInterchange/profitability.ll:
```diff
@@ -174,23 +174,15 @@ for.end12:
 ;;     for(int j=1;j<100;j++)
 ;;       A[j][0] = A[j][0] + B[j][i];
+;
+; FIXME
```
Yep, thanks, will fix.

sjoerdmeijer reviewed (commented):
> The implementation looks good to me. but some tests do not appear to have achieved their original objectives. I'm not sure what we should do in such cases, but in my opinion a mechanical update does not make sense. Marking `XFAIL` may be one option.

Thanks, and I also had a bit of a nagging feeling about the test cases. I quite like the idea about the XFAILs. I think what I will do is the following:
- I will first convert the test cases that were modified and check the full IR to tests that check the optimisation remark in a separate patch. Checking the the optimisation remark is more compact, less verbose and easier to see. 
- I will then see if adding XFAILS here is better, I think so.

kasuga-fj reviewed (commented):
> Thanks, and I also had a bit of a nagging feeling about the test cases. I quite like the idea about the XFAILs. I think what I will do is the following:
>
> - I will first convert the test cases that were modified and check the full IR to tests that check the optimisation remark in a separate patch. Checking the the optimisation remark is more compact, less verbose and easier to see.
>- I will then see if adding XFAILS here is better, I think so.

Cool! I second this idea.

sjoerdmeijer committed:
[LoopInterchange] Remove "S" Scalar Dependencies

We are not handling 'S' scalar dependencies correctly and have at least
the following miscompiles related to that:

[LoopInterchange] incorrect handling of scalar dependencies and dependence vectors starting with ">" #54176
[LoopInterchange] Interchange breaks program correctness #46867
[LoopInterchange] Loops should not interchanged due to dependencies #47259
[LoopInterchange] Loops should not interchanged due to control flow #47401

This patch does no longer insert the "S" dependency/direction into the
dependency matrix, so a dependency is never "S". We seem to have
forgotten what the exact meaning is of this dependency type, and don't
see why it should be treated differently.

We prefer correctness over incorrect and more aggressive results. I.e.,
this prevents the miscompiles at the expense of handling less cases,
i.e. making interchange more pessimistic. However, some of the cases
that are now rejected for dependence analysis reasons, were rejected
before too but for other reasons (e.g. profitability). So at least for
the llvm regression tests, the number of regression are very reasonable.
This should be a stopgap. We would like to get interchange enabled by
default and thus prefer correctness over unsafe transforms, and later
see if we can get solve the regressions.

sjoerdmeijer commented:
I have now XFAIL'ed the test cases, with 2 exceptions where it was already rejected but for different reasons.

kasuga-fj commented on llvm/test/Transforms/LoopInterchange/lcssa.ll:
```diff
@@ -177,7 +177,7 @@ for.end16:                                        ; preds = %for.exit
 
 ; PHI node in inner latch with multiple predecessors.
-; REMARK: Interchanged
+; REMARK:      Dependence
```
Should this also be XFAIL?

sjoerdmeijer replied:
Fixed

kasuga-fj reviewed (approved):
Thanks, LGTM, now all that remains is to apply the formatter

sjoerdmeijer commented:
> Thanks, LGTM, now all that remains is to apply the formatter

Thanks a lot for the reviews, will do that before merging this.

Meinersbur reviewed (approved):
LGTM
